### PR TITLE
Enhance `predict` API to serve for env validation purpose.

### PR DESCRIPTION
**Pull request by B-Step62**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/B-Step62/mlflow/pull/10759?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/10759/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 10759
```

</p>
</details>

### What changes are proposed in this pull request?

Enhance `mlflow models predict` CLI command so that users can use it (more easily) for validating model environment before deployment.
1. Add Python API for convenience in the notebook environment. Almost same as CLI, but also support serialized json/csv string as input data (while CLI only supports file path or stdin) for the sake of notebook experience.
2. Show better error message and guidance when some packages are missing in the model. Basically introduce usage of `extra_pip_requirements`.
3. Add `pip-requirement-override` argument to both Python/CLI APIs, so that users can try additional/updated dependencies without having to create and log model (as Python only emits error one by one, this process can be very annoying otherwise).

**Note**
Will work on OSS/Databricks docs as a follow-up.

### How is this PR tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [x] Manual tests

Verified the functionality works in Databricks. Note that it takes a bit log for the first time creating virtualenv (0:20~0:45 in the video).

https://github.com/mlflow/mlflow/assets/31463517/c26dd903-2954-45df-8440-042f76c291ec

In Databricks, I was able to test only with virtualenv as we don't have conda installed. For conda, I tested on devbox and it worked as same.

### Does this PR require documentation update?
Will work on doc update.

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Enhanced `predict` API for MLflow Model so it can be used for inference environment validation before model deployment: (1) add Python API for the sake of notebook convenience (2) introduce `pip-requirement-overrides` argument to test dependency change (3) enrich error message.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/pyfunc/backend.py [hunk 1]
```diff
@@ -161,8 +170,31 @@ def predict(self, model_uri, input_path, output_path, content_type):
...
+                ]
+
+            environment = self.prepare_env(
+                local_path, pip_requirements_override=pip_requirements_override
+            )
+
+            try:
+                environment.execute(" ".join(predict_cmd))
+            except ShellCommandException as e:
+                raise ShellCommandException(
```
Comment (B-Step62) [inline mlflow/pyfunc/backend.py:187]: Just to show additional messages so that users won't overlook the subprocess error shown above the main process error.

File: mlflow/models/__init__.py [hunk 2]
```diff
@@ -86,6 +89,64 @@ def build_docker(
     )
 
 
+def predict(
+    model_uri: str,
+    input_data_or_path: str,
+    output_path: Optional[str] = None,
```
Comment (dbczumar) [inline mlflow/models/__init__.py:95]: As discussed offline, would also be awesome to accept Python dict / list

File: mlflow/models/__init__.py [hunk 3]
```diff
@@ -86,6 +89,64 @@ def build_docker(
...
+    :param content_type: Content type of the input file. Can be one of {‘json’, ‘csv’}.
+    :param env_manager: If specified, create an environment for MLmodel using the specified
+        environment manager. The following values are supported:
+            - virtualenv (default): use virtualenv (and pyenv for Python version management)
+            - local: use the local environment
+            - conda: use conda
+    :param install_mlflow: If specified and there is a conda or virtualenv environment to be
+        activated mlflow will be installed into the environment after it has been activated.
+        The version of installed mlflow will be the same as the one used to invoke this command.
+    :param pip_requirements_override: If specified, install the specified python dependencies to
```
Comment (dbczumar) [inline mlflow/models/__init__.py:119]: Can we add an example for this in the docstring?

File: mlflow/models/__init__.py [hunk 4]
```diff
@@ -86,6 +89,64 @@ def build_docker(
...
+        the model inference environment. This is particularly useful when you want to add extra
+        dependencies or try different versions of the dependencies defined in the logged model.
+    """
+    if not isinstance(input_data_or_path, str):
+        raise TypeError(
```
Comment (dbczumar) [inline mlflow/models/__init__.py:124]: Nit: Prefer to raise `MlflowException.invalid_parameter_value()`

File: mlflow/pyfunc/_mlflow_pyfunc_backend_predict.py [hunk 5]
```diff
@@ -15,14 +16,43 @@ def parse_args():
...
+        _predict(
+            model_uri=args.model_uri,
+            input_path=args.input_path if args.input_path else None,
+            output_path=args.output_path if args.output_path else None,
+            content_type=args.content_type,
+        )
+    except ModuleNotFoundError as e:
+        missing_module = re.search(r"No module named '(.*)'", str(e)).group(1)
+        message = _MISSING_MODULE_HELP_MSG.format(e=str(e), missing_module=missing_module)
+        raise RuntimeError(message) from e
```
Comment (dbczumar) [inline mlflow/pyfunc/_mlflow_pyfunc_backend_predict.py:55]: What does re-raising RuntimeError do?

File: mlflow/pyfunc/backend.py [hunk 6]
```diff
@@ -161,8 +170,31 @@ def predict(self, model_uri, input_path, output_path, content_type):
...
+                    f"{e}\n\nAn exception occurred while running model prediction within a "
+                    f"{self._env_manager} environment. You can find the error message "
+                    f"from the prediction subprocess by scrolling above."
+                ) from None
         else:
+            if pip_requirements_override:
+                raise Exception(
```
Comment (dbczumar) [inline mlflow/pyfunc/backend.py:194]: Nit: Can we raise MlflowExceptions where possible?

File: mlflow/models/__init__.py [hunk 7]
```diff
@@ -86,6 +89,64 @@ def build_docker(
...
 
 
+def predict(
+    model_uri: str,
+    input_data_or_path: str,
+    output_path: Optional[str] = None,
+    content_type: str = "json",
+    env_manager: _EnvManager = _EnvManager.VIRTUALENV,
+    install_mlflow: bool = False,
+    pip_requirements_override: Optional[List[str]] = None,
```
Comment (dbczumar) [inline mlflow/models/__init__.py:99]: When `predict()` throws an error related to requirements, can we include info about how to use this param to test a different env. in the error message? Can we also include info about how to update a model's dependencies?

File: mlflow/pyfunc/_mlflow_pyfunc_backend_predict.py (hunk 5 above)
Comment (B-Step62) [reply mlflow/pyfunc/_mlflow_pyfunc_backend_predict.py:55]: Just to show additional messages (and at the bottom of the stack trace rather than log.error)

File: mlflow/models/__init__.py (hunk 7 above)
Comment (B-Step62) [reply mlflow/models/__init__.py:99]: Makes sense, what about adding a small section/FAQ in [the new deployment doc](https://output.circle-artifacts.com/output/job/3b0d72cf-8085-4437-80a0-0cf1bc0e583e/artifacts/0/docs/build/html/deployment/index.html#faq) (which is already merged and will be deployed soon) and show link to the section in the error message?

Comment (dbczumar) [reply mlflow/models/__init__.py:99]: Love it!

File: mlflow/models/python_api.py [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+from io import StringIO
+from typing import Any, Dict, List, Optional, Union
+
+from mlflow.exceptions import MlflowException
+from mlflow.models.flavor_backend_registry import get_flavor_backend
+from mlflow.utils import env_manager as _EnvManager
+from mlflow.utils.file_utils import TempDir
+
+
+def build_docker(
```
Comment (B-Step62) [inline mlflow/models/python_api.py:12]: No change from the original definition in `mlflow/models/__init__.py`. Just moved to avoid complicating `__init__.py`.

File: mlflow/models/__init__.py [hunk 9]
```diff
@@ -105,6 +63,7 @@ def build_docker(
 # Under skinny-mlflow requirements, the following packages cannot be imported
 # because of lack of numpy/pandas library, so wrap them with try...except block
 try:
+    from mlflow.models.python_api import predict
```
Comment (B-Step62) [inline mlflow/models/__init__.py:66]: `predict` needs pandas but `build-docker` is not, which needs a bit dirty solution in `predict` impl. Wondering if we really need to make `build-docker` available in mlflow-skinny.

File: mlflow/models/python_api.py [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+
+_CONTENT_TYPE_CSV = "csv"
+_CONTENT_TYPE_JSON = "json"
+
+
+def predict(
+    model_uri: str,
+    # TODO: This is currently subset of PyfuncInput, ideally we should cover all
+    input_data: Union[str, Dict[str, Any], List[Any], "pd.DataFrame", None] = None,  # noqa: F821
+    input_path: Optional[str] = None,
```
Comment (B-Step62) [inline mlflow/models/python_api.py:59]: Separated input_path and input_data arguments, as pandas `read_csv` combined with StringIO is too permissive. For exampl,e `pd.read_csv(StringIO("some-incorrect-file-path.csv"))` will be read as single column DF (while we want to say file not exists).

File: mlflow/models/python_api.py [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+            output_path=output_path,
+            content_type=content_type,
+            pip_requirements_override=pip_requirements_override,
+        )
+
+    if input_data is not None and input_path is not None:
+        raise MlflowException.invalid_parameter_value(
+            "Both input_data and input_path are provided. Only one of them should be specified."
+        )
+    elif input_data is not None:
```
Comment (B-Step62) [inline mlflow/models/python_api.py:141]: Not `input_data` as this can be pandas dataframe, which doesn't allow casting to single boolean.

File: mlflow/utils/cli_args.py [hunk 12]
```diff
@@ -83,7 +83,7 @@ def _resolve_env_manager(_, __, env_manager):
 def _create_env_manager_option(help_string):
     return click.option(
         "--env-manager",
-        default=None,
+        default=_EnvManager.VIRTUALENV,
```
Comment (B-Step62) [inline mlflow/utils/cli_args.py:86]: Every cli command uses this default (by doing `env_manager = env_manager or _EnvMangager.VIRTUALENV`)

File: mlflow/utils/cli_args.py [hunk 13]
```diff
@@ -80,17 +80,18 @@ def _resolve_env_manager(_, __, env_manager):
...
-        default=None,
+        default=default,
         type=click.UNPROCESSED,
         callback=_resolve_env_manager,
         help=help_string,
     )
 
 
 ENV_MANAGER = _create_env_manager_option(
+    default=_EnvManager.VIRTUALENV,
```
Comment (B-Step62) [inline mlflow/utils/cli_args.py:94]: Every cli command was using this default (by doing `env_manager = env_manager or _EnvMangager.VIRTUALENV`)

File: mlflow/pyfunc/_mlflow_pyfunc_backend_predict.py [hunk 14]
```diff
@@ -15,14 +16,46 @@ def parse_args():
     return parser.parse_args()
 
 
+# Guidance for fixing missing module error
+_MISSING_MODULE_HELP_MSG = (
+    "Exception occurred while running inference: {e}"
+    "\n\n"
+    "\033[93m[Hint] It appears to be your MLflow Model doesn't contain the required "
```
Comment (BenWilson2) [inline mlflow/pyfunc/_mlflow_pyfunc_backend_predict.py:23]:
```suggestion
    "\033[93m[Hint] It appears that your MLflow Model doesn't contain the required "
```

File: mlflow/pyfunc/_mlflow_pyfunc_backend_predict.py [hunk 15]
```diff
@@ -15,14 +16,46 @@ def parse_args():
...
+    "dependency '{missing_module}' to run model inference. When logging a model, MLflow "
+    "detects dependencies based on the model flavor but it is possible that some "
```
Comment (BenWilson2) [inline mlflow/pyfunc/_mlflow_pyfunc_backend_predict.py:25]:
```suggestion
    "detects dependencies based on the model flavor, but it is possible that some "
```

**Review by BenWilson2 — APPROVED** [review]
Nice implementation!

File: mlflow/models/python_api.py (hunk 10 above)
Comment (BenWilson2) [reply mlflow/models/python_api.py:59]: Good idea; the API makes sense and is clear.

File: docs/source/deployment/index.rst [hunk 16]
```diff
@@ -169,3 +169,144 @@ Almost all functionalities available in MLflow deployment can also be accessed v
...
+
+How to test my model before deploying to the production environment?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Testing your model before deployment is a critical step to ensure production readiness.
+MLflow provides a few ways to test your model locally, either in a virtual environment or a Docker container.
+
+Testing offline prediction with a virtual environment
+*****************************************************
+You can use MLflow `predict` API via CLI or Python to making test predictions with your model.
+This wiill load your model from the model URI, create a virtual environemnt with the model dependencies (defined in MLflow Model),
```
Comment (dbczumar) [inline docs/source/deployment/index.rst:181]:
```suggestion
This will load your model from the model URI, create a virtual environment with the model dependencies (defined in MLflow Model),
```

File: docs/source/deployment/index.rst [hunk 17]
```diff
@@ -169,3 +169,144 @@ Almost all functionalities available in MLflow deployment can also be accessed v
...
+    .. code-tab:: python
+
+        import mlflow
+
+        mlflow.model.predict(
+            model_uri="runs:/<run_id>/model",
+            input_data=<input_data>,
+        )
+
+Using the `predict` API is convenient for testing your model and inference environment as being a relatively light weight and quick.
```
Comment (dbczumar) [inline docs/source/deployment/index.rst:200]:
```suggestion
Using the `predict` API is convenient for testing your model and inference environment quickly.
```

File: docs/source/deployment/index.rst [hunk 18]
```diff
@@ -169,3 +169,144 @@ Almost all functionalities available in MLflow deployment can also be accessed v
...
+
+    .. code-tab:: bash
+
+        mlflow models predict -m runs:/<run_id>/model-i <input_path>
+
+    .. code-tab:: python
+
+        import mlflow
+
+        mlflow.model.predict(
```
Comment (dbczumar) [inline docs/source/deployment/index.rst:195]: This should be `mlflow.models` instead of `mlflow.model`

File: docs/source/deployment/index.rst [hunk 19]
```diff
@@ -169,3 +169,144 @@ Almost all functionalities available in MLflow deployment can also be accessed v
...
+    curl -X POST -H "Content-Type: application/json" \
+        --data '{"inputs": [[1, 2], [3, 4]]}' \
+        http://localhost:<port>/invocations
+
+
+While this is reliable way to test your model before deployment, one caveat is that virtual environment doesn't absorb the OS level differences
+between your machine and the production environment. For example, if you are using Windows as a local dev machine but your deployment target is
+running on Linux, you may encounter some issues that are not reproducible in the virtual environment.
+
+In this case, you can use Docker container to test your model.
```
Comment (dbczumar) [inline docs/source/deployment/index.rst:224]: I don't think you can run Windows docker containers on Linux machines, right?

File: docs/source/deployment/index.rst [hunk 20]
```diff
@@ -169,3 +169,144 @@ Almost all functionalities available in MLflow deployment can also be accessed v
...
+        # If you want to define all dependencies from scratch, use `pip_requirements` option.
+        # Both options also accept a path to a pip requirements file e.g. requirements.txt.
+        extra_pip_requirements=["opencv-python==4.8.0"],
+    )
+
+.. note::
+
+    Alternatively, you can directly update the model metadata stored in the artifact storage. For example, `<path-to-model>/requirements.txt`
+    defines pip dependencies to be installed for serving your model (or `<path-to-model>/conda.yaml` if you use conda as a package manager).
+    However, this approach is not recommended because this is error-prone and also you need to do this for every model version.
```
Comment (dbczumar) [inline docs/source/deployment/index.rst:312]: I don't think it's error prone if we give users good code to do this. Can we provide a snippet here? Maybe this deserves its own API (good to track with a follow-up JIRA).

File: docs/source/deployment/index.rst [hunk 21]
```diff
@@ -169,3 +169,144 @@ Almost all functionalities available in MLflow deployment can also be accessed v
...
+            pip_requirements="opencv-python==4.8.0",
+        )
+
+The specified dependencies will be installed to the virtual environment in addition to (or instead of) the dependencies
+defined in the model metadata. Since this doesn't mutate the model, you can iterate quickly and safely to find the correct dependencies.
+
+3. Update the model metadata
+****************************
+Once you find the correct dependencies, you can create a new model with the correct dependencies.
+To do so, specify either `pip_requirements` or `extra_pip_requirements` option when logging the model.
```
Comment (dbczumar) [inline docs/source/deployment/index.rst:294]: Specifying `pip_requirements` will get rid of the other requirements that MLflow infers, right? I think users always want `extra_pip_requirements` to solve missing dependency issues.

File: mlflow/models/__init__.py (hunk 9 above)
Comment (dbczumar) [reply mlflow/models/__init__.py:66]: We probably don't, but best not to drop it now...thanks for handling this

File: mlflow/models/python_api.py [hunk 22]
```diff
@@ -0,0 +1,200 @@
...
+    )
+
+
+_CONTENT_TYPE_CSV = "csv"
+_CONTENT_TYPE_JSON = "json"
+
+
+def predict(
+    model_uri: str,
+    # TODO: This is currently subset of PyfuncInput, ideally we should cover all
```
Comment (dbczumar) [inline mlflow/models/python_api.py:57]: Can we file a ticket to address this and prioritize as a fast-follow? This is pretty important :)

File: tests/models/test_python_api.py [hunk 23]
```diff
@@ -0,0 +1,180 @@
+import json
```
Comment (dbczumar) [inline tests/models/test_python_api.py:1]: @B-Step62 Can we add test coverage for `pip_requirements_override` with `conda` and `virtualenv` as env managers? Looks like all the cases in this suite set `pip_requirements_override` to `None`

File: docs/source/deployment/index.rst (hunk 19 above)
Comment (B-Step62) [reply docs/source/deployment/index.rst:224]: Right but this one talks about Linux on Windows (which I think much more common the the other)

File: docs/source/deployment/index.rst (hunk 20 above)
Comment (B-Step62) [reply docs/source/deployment/index.rst:312]: Ahh sorry here I mean manual file editting e.g. ssh&vi, let me rephrase it. Agree with having programatic way is good so will create a follow-up JIRA!

File: mlflow/models/python_api.py (hunk 22 above)
Comment (B-Step62) [reply mlflow/models/python_api.py:57]: Sure! Created and added to next sprint.

File: tests/models/test_python_api.py (hunk 23 above)
Comment (B-Step62) [reply tests/models/test_python_api.py:1]: @dbczumar sure I will create a follow-up PR for that!

File: docs/source/deployment/index.rst (hunk 19 above)
Comment (dbczumar) [reply docs/source/deployment/index.rst:224]: I think we're positioning Docker as a general solution for cross-OS testing, which isn't always true. Should we clarify?

File: tests/models/test_python_api.py (hunk 23 above)
Comment (dbczumar) [reply tests/models/test_python_api.py:1]: Can we add it as part of this PR? Else, how do we know it's safe to merge this PR?

File: docs/source/deployment/index.rst (hunk 19 above)
Comment (B-Step62) [reply docs/source/deployment/index.rst:224]: True, will add some clarification that Docker is not designed for full cross-OS capability while being a light weight testing option for some cases.

File: tests/models/test_python_api.py (hunk 23 above)
Comment (B-Step62) [reply tests/models/test_python_api.py:1]: I'm fine with doing it here, just tried to make it easy to review. (Btw I tested the functionality for those env managers  in manual E2E test above).

File: mlflow/pyfunc/_mlflow_pyfunc_backend_predict.py [hunk 24]
```diff
@@ -15,14 +16,46 @@ def parse_args():
...
+
+    try:
+        _predict(
+            model_uri=args.model_uri,
+            input_path=args.input_path if args.input_path else None,
+            output_path=args.output_path if args.output_path else None,
+            content_type=args.content_type,
+        )
+    except ModuleNotFoundError as e:
+        missing_module = re.search(r"No module named '(.*)'", str(e)).group(1)
```
Comment (harupy) [inline mlflow/pyfunc/_mlflow_pyfunc_backend_predict.py:56]: /ai Does `ModuleNotFoundError` have an attribute for the name of the module that's not found?

Comment (harupy) [inline mlflow/pyfunc/_mlflow_pyfunc_backend_predict.py:56]: Does `ModuleNotFoundError` have an attribute for the module name?

Comment (harupy) [reply mlflow/pyfunc/_mlflow_pyfunc_backend_predict.py:56]: Example

```python
try:
    import foobar

except ModuleNotFoundError as e:
    print(e.name)  # -> foobar

```

Comment (B-Step62) [reply mlflow/pyfunc/_mlflow_pyfunc_backend_predict.py:56]: Oh it seems `e.name` works here, great finding!

File: mlflow/models/cli.py [hunk 25]
```diff
@@ -125,28 +125,20 @@ def serve(
...
-    install_mlflow,
-):
+@click.option(
+    "--pip-requirements-override",
+    "-r",
+    default=None,
+    help="Specify packages and versions to override the dependencies defined"
+    "in the model. Must be a comma-separated string like x==y,z==a.",
+)
+def predict(**kwargs):
```
Comment (harupy) [inline mlflow/models/cli.py:135]: curious why we need to use `kwargs`?

Comment (B-Step62) [reply mlflow/models/cli.py:135]: To make it explicit that the Python API and CLI are same (and they should be)

File: mlflow/models/cli.py [hunk 26]
```diff
@@ -125,28 +125,20 @@ def serve(
...
-    content_type,
-    env_manager,
-    install_mlflow,
-):
+@click.option(
+    "--pip-requirements-override",
+    "-r",
+    default=None,
+    help="Specify packages and versions to override the dependencies defined"
+    "in the model. Must be a comma-separated string like x==y,z==a.",
```
Comment (harupy) [inline mlflow/models/cli.py:133]:
```suggestion
    help="Specify packages and versions to override the dependencies defined "
    "in the model. Must be a comma-separated string like x==y,z==a.",
```

File: mlflow/models/cli.py (hunk 25 above)
Comment (harupy) [reply mlflow/models/cli.py:135]: It looks like this is the simplest way. LGTM

File: docs/source/deployment/index.rst [hunk 27]
```diff
@@ -169,3 +169,144 @@ Almost all functionalities available in MLflow deployment can also be accessed v
...
 ---
+
+How to test my model before deploying to the production environment?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Testing your model before deployment is a critical step to ensure production readiness.
+MLflow provides a few ways to test your model locally, either in a virtual environment or a Docker container.
+
+Testing offline prediction with a virtual environment
+*****************************************************
+You can use MLflow `predict` API via CLI or Python to making test predictions with your model.
```
Comment (harupy) [inline docs/source/deployment/index.rst:180]:
```suggestion
You can use `mlflow models predict` API via CLI or Python to making test predictions with your model.
```

Perhaps it's just me, but `predict` API sounds ambiguous. Can we use the full command name?

Comment (B-Step62) [reply docs/source/deployment/index.rst:180]: Makes sense to me, will replace:)

**Review by harupy — APPROVED** [review]
LGTM :)
